common/types.go

```diff
@@ -20,6 +20,55 @@ func NewBigInt(v int64) BigInt {
 	return BigInt{*big.NewInt(v)}
 }
 
+func (b BigInt) Plus(other BigInt) BigInt {
+	result := big.Int{}
+	result.Set(&b.Int) // creates a copy of b.Int
+	result.Add(&result, &other.Int)
+	return BigInt{result}
+}
+
+func (b BigInt) Times(other BigInt) BigInt {
+	result := big.Int{}
+	result.Set(&b.Int) // creates a copy of b.Int
+	result.Mul(&result, &other.Int)
+	return BigInt{result}
+}
+
+func (b BigInt) MarshalText() ([]byte, error) {
+	return b.Int.MarshalText()
+}
+
+func (b *BigInt) UnmarshalText(text []byte) error {
+	return b.Int.UnmarshalText(text)
+}
+
+// Used by the `cbor` library.
+func (b BigInt) MarshalBinary() ([]byte, error) {
```

Comment:
Wow indeed weird. It looks like this is due to cbor's internal handling of BigInt marsahaling and not detecting it whenever it's embedded in another struct. 

Would it be safer to use the cbor's library internal marshal/unmarshal methods for big ints? Seem a bit more complex (but didn't look in depth) e.g.:
https://github.com/fxamacker/cbor/blob/a1d4c29efb5eac2ad571787a350fc6b9c46995b1/encode.go#L1179-L1212


One way to achieve this would be to do:
```
type BigInt struct {
	big.Int `cbor:"int"`
}
```
with the added cbor tag it seems to work.

Reply:
Interesting. I don't know that it's safer, but it would be more interoperable/standard :/. Though it'd better to go with a more efficient encoding:
```go
type BigInt struct {
	_   struct{} `cbor:",toarray"`
	Int big.Int
}
```
so we're not encoding the map structure over and over.

FWIW, nothing special is needed for `cbor` to encode the `big.Int` using the special encoder. It's enough even just to make the field explicit:
```go
type BigInt struct {
	Int big.Int
}
```
In this case, it encodes a large number as
```
A1                                      # map(1)
   63                                   # text(3)
      496E74                            # "Int"
   C2                                   # tag(2)
      50                                # bytes(16)
         3FFFFFFFFFFFFFFF0000000000000001 # "?\xFF\xFF\xFF\xFF\xFF\xFF\xFF\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0001"
```
Values smaller than MaxInt64 were roundtripped correctly, but the large numbers in the test errored out during decoding with "cbor: CBOR tag isn't allowed".
This is because nexus (re)uses the oasis-core `cbor` wrapper for max compatibility, and that wrapper disallows tags by default: https://github.com/oasisprotocol/oasis-core/blob/799668903b5a993d0f8af0110c99f2916fb6eb11/go/common/cbor/cbor.go#L43

So our options are:
- Leave PR as-is. Downside: Not a universally understood encoding; you need our BigInt definition to decode.
- Create our own CBOR utility funcs that allow BigInt tags, do not use oasis-core's utilities. Downside: Boilerplate, potential (though unlikely) tiny discrepancies with oasis-core behavior.

I prefer the first option largely because it's less work at this point, and because this will be used internally only. But open to stronger opinions.

Reply:
I would leave the PR as-is, the encoding seems similar to `Quantity` (except for the extra sign) and we do not really use CBOR tags elsewhere.